docs/website/docs/reference/bindings/c-api.md

```diff
@@ -1,216 +1,353 @@
+    custom targets, dialects, etc.
+    Applications interface with the compiler shared object through the compiler
+    C API's exported symbols.
+  }
 
-By default, IREE uses its own tiny Virtual Machine (VM) at runtime to interpret
-program instructions on the host system. VM instructions may also be lowered
-further to LLVM IR, C, or other representations for static or resource
-constrained deployment.
+  subgraph compiler[libIREECompiler.so / IREECompiler.dll]
+    pipelines("Pipelines
 
-The VM supports generic operations like loads, stores, arithmetic, function
-calls, and control flow. It builds streams of more complex program logic and
-dense math into command buffers that are dispatched to hardware backends
-through the Hardware Abstraction Layer (HAL) interface.
+    • Flow
+    • Stream
+    • etc.")
 
-Most interaction with IREE's C API involves either the VM or the HAL.
+    targets("Target backends
 
-<!-- TODO(scotttodd): diagrams -->
+    • llvm-cpu
+    • vulkan-spirv
+    • etc.")
 
-### IREE VM
+    passes("General passes
 
-* VM _instances_ can serve multiple isolated execution _contexts_
-* VM _contexts_ are effectively sandboxes for loading modules and running
-  programs
-* VM _modules_ provide extra functionality to execution _contexts_, such as
-  access to hardware accelerators through the HAL. Compiled user programs are
-  also modules.
+    • Const eval
+    • DCE
+    • etc.")
+  end
 
-### IREE HAL
+  plugins("Compiler plugins
 
-* HAL _drivers_ are used to enumerate and create HAL _devices_
-* HAL _devices_ interface with hardware, such as by allocating device memory,
-  preparing executables, recording and dispatching command buffers, and
-  synchronizing with the host
-* HAL _buffers_ and _buffer views_ represent storage and shaped/typed views
-  into that storage (aka "tensors")
+    • Custom targets
+    • Custom dialects
+    • etc.")
 
-## Using the C API
+  application(Your application)
 
-### Setup
+  compiler <-- "Plugin API<br>(static or dynamic linking)" --> plugins
+  compiler -. "Compiler C API<br>(exported symbols)" .-> application
+```
 
-Include headers:
+API definitions can be found in the following locations:
+
+| Source location | Overview |
+| --------------- | -------- |
+[`iree/compiler/embedding_api.h`](https://github.com/openxla/iree/blob/main/compiler/bindings/c/iree/compiler/embedding_api.h) | Top-level IREE compiler embedding API
+[`iree/compiler/PluginAPI/` directory](https://github.com/openxla/iree/tree/main/compiler/src/iree/compiler/PluginAPI) | IREE compiler plugin API
+[`mlir/include/mlir-c/` directory](https://github.com/llvm/llvm-project/tree/main/mlir/include/mlir-c) | MLIR C API headers
+
+### Concepts
+
+The compiler API is centered around running pipelines to translate inputs to
+artifacts. These are modeled via _sessions_, _invocations_, _sources_, and
+_outputs_.
+
+```mermaid
+stateDiagram-v2
+  accTitle: IREE compiler session and invocation state diagram
```

Comment:
This diagram is currently set up to show the many-to-many relationship between inputs, invocations, and outputs:
![image](https://github.com/openxla/iree/assets/4010439/a284a37a-f9ba-4684-bcea-abcd426d554a)

I could simplify it to just show one path through:
![image](https://github.com/openxla/iree/assets/4010439/dc9c8c26-8766-469f-800c-7970fed79d22)